main/src/library/File.flix

```diff
@@ -840,4 +840,103 @@ namespace File {
+                Err(getMessage(ex))
+        }
+
+    ///
+    /// Returns an iterator naming the files and directories in the directory `f`.
+    ///
+    /// Does not recursively traverse the directory.
+    ///
+    pub def listWithIter(f: String): Result[Iterator[String], String] & Impure =
+        try {
+            import new java.io.File(String): ##java.io.File & Impure as newFile;
+            import java.io.File.toPath(): ##java.nio.file.Path & Impure;
+            import static java.nio.file.Files.newDirectoryStream(##java.nio.file.Path): ##java.nio.file.DirectoryStream & Impure;
+            import java.nio.file.DirectoryStream.iterator(): ##java.util.Iterator & Impure;
+
+            import java.util.Iterator.hasNext(): Bool & Impure;
+            import java.util.Iterator.next(): ##java.nio.file.Path & Impure as getNext;
+
+            import java.nio.file.Path.getFileName(): ##java.nio.file.Path & Impure;
+            import java.nio.file.Path.toString(): String & Impure;
+
+            let javaFile = newFile(f);
+            let javaPath = toPath(javaFile);
+            let javaStream = newDirectoryStream(javaPath);
+            let javaIter = iterator(javaStream);
+
+            let done = () -> not hasNext(javaIter);
+            let next = () -> {
+                if(hasNext(javaIter)) {
+                    let name = getFileName(getNext(javaIter));
+                    toString(name)
+                } else {
+                    bug!("Empty iterator.")
+                }
+            };
+
+            Ok(Iterator(done, next))
+        } catch {
+            case ex: ##java.io.IOException =>
+                import java.lang.Throwable.getMessage(): String & Impure;
+                Err(getMessage(ex))
+        }
+
+    ///
+    /// Returns an iterator naming the files and directories in the directory `f`.
+    ///
+    /// Recursively traverses the directory.
+    ///
+    pub def walk(f: String): Result[Iterator[String], String] & Impure =
+        try {
+            import new java.io.File(String): ##java.io.File & Impure as newFile;
+            import java.io.File.toPath(): ##java.nio.file.Path & Impure;
+            import static java.nio.file.Files.walk(##java.nio.file.Path, Array[##java.nio.file.FileVisitOption]): ##java.util.stream.Stream & Impure;
+            import java.util.stream.BaseStream.iterator(): ##java.util.Iterator & Impure;
+
+            import java.util.Iterator.hasNext(): Bool & Impure;
+            import java.util.Iterator.next(): ##java.nio.file.Path & Impure as getNext;
+
+            import java.nio.file.Path.toString(): String & Impure as pathToString;
+
+            let javaFile = newFile(f);
+            let javaPath = toPath(javaFile);
+            let javaStream = walk(javaPath, []);
+            let javaIter = iterator(javaStream as ##java.util.stream.BaseStream);
+
+            let done = () -> not hasNext(javaIter);
```

Comment:
I see some duplication. How about introducing a private helper function at the bottom which converts a Java iterator to a Flix Iterator?